components/style/values/distance.rs

```diff
@@ -0,0 +1,124 @@
+    Sqrt(f64),
+    /// Represented as the squared distance itself.
+    Value(f64),
+}
+
+impl ComputeSquaredDistance for u16 {
+    #[inline]
+    fn compute_squared_distance(&self, other: &Self) -> Result<SquaredDistance, ()> {
+        Ok(SquaredDistance::Sqrt(((*self as f64) - (*other as f64)).abs()))
+    }
+}
+
+impl ComputeSquaredDistance for i32 {
+    #[inline]
+    fn compute_squared_distance(&self, other: &Self) -> Result<SquaredDistance, ()> {
+        Ok(SquaredDistance::Sqrt((*self - *other).abs() as f64))
+    }
+}
+
+impl ComputeSquaredDistance for f32 {
+    #[inline]
+    fn compute_squared_distance(&self, other: &Self) -> Result<SquaredDistance, ()> {
+        Ok(SquaredDistance::Sqrt((*self - *other).abs() as f64))
+    }
+}
+
+impl ComputeSquaredDistance for f64 {
+    #[inline]
+    fn compute_squared_distance(&self, other: &Self) -> Result<SquaredDistance, ()> {
+        Ok(SquaredDistance::Sqrt((*self - *other).abs()))
+    }
+}
+
+impl ComputeSquaredDistance for Au {
+    #[inline]
+    fn compute_squared_distance(&self, other: &Self) -> Result<SquaredDistance, ()> {
+        self.0.compute_squared_distance(&other.0)
+    }
+}
+
+impl<T> ComputeSquaredDistance for Option<T>
+    where T: ComputeSquaredDistance
+{
+    #[inline]
+    fn compute_squared_distance(&self, other: &Self) -> Result<SquaredDistance, ()> {
+        match (self.as_ref(), other.as_ref()) {
+            (Some(this), Some(other)) => this.compute_squared_distance(other),
+            (None, None) => Ok(SquaredDistance::Value(0.)),
+            _ => Err(()),
+        }
+    }
+}
+
+impl<T> ComputeSquaredDistance for Size2D<T>
+    where T: ComputeSquaredDistance
+{
+    #[inline]
+    fn compute_squared_distance(&self, other: &Self) -> Result<SquaredDistance, ()> {
+        Ok(self.width.compute_squared_distance(&other.width)? + self.height.compute_squared_distance(&other.height)?)
+    }
+}
+
+impl SquaredDistance {
+    /// Returns the square root of this squared distance.
+    pub fn sqrt(self) -> f64 {
+        match self {
+            SquaredDistance::Sqrt(this) => this,
+            SquaredDistance::Value(this) => this.sqrt(),
+        }
+    }
+}
+
+impl From<SquaredDistance> for f64 {
+    #[inline]
+    fn from(distance: SquaredDistance) -> Self {
+        match distance {
+            SquaredDistance::Sqrt(this) => this * this,
+            SquaredDistance::Value(this) => this,
+        }
+    }
+}
+
+impl Add for SquaredDistance {
+    type Output = Self;
+
+    #[inline]
+    fn add(self, rhs: Self) -> Self {
+        SquaredDistance::Value(f64::from(self) + f64::from(rhs))
+    }
+}
+
+impl Sum for SquaredDistance {
+    fn sum<I>(mut iter: I) -> Self
+    where
+        I: Iterator<Item = Self>,
+    {
+        let first = match iter.next() {
+            Some(first) => first,
+            None => return SquaredDistance::Value(0.),
+        };
+        iter.fold(first, Add::add)
```

Comment:
I'm not doing that to avoid pointlessly transforming a `SquaredDistance::Sqrt` into `SquaredDistance::Value` if there was only one distance in the iterator. As I said in a comment of the PR though, I'm not sure we should even bother to try to avoid the final `f64::sqrt` call, I just did it that way to preserve the functionality but I wouldn't mind just storing the squared one always.